File: src/623. Add One Row to Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# DFS
# class Solution:
#     def addOneRow(self, root: Optional[TreeNode], val: int, depth: int) -> Optional[TreeNode]:
#         if depth == 1: return TreeNode(val, root, None)
#         self.graft(root, val, depth)
#         return root
#     def graft(self, root: Optional[TreeNode], val: int, depth: int) -> None:
#         if not root: return
#         if depth == 2:
#             root.left = TreeNode(val, root.left, None)
#             root.right = TreeNode(val, None, root.right)
#             return
#         self.graft(root.left, val, depth - 1)
#         self.graft(root.right, val, depth - 1)

# BFS
class Solution:
    def addOneRow(self, root: Optional[TreeNode], val: int, depth: int) -> Optional[TreeNode]:
        dummy = TreeNode(None, root, None)
        queue = [dummy]
        for _ in range(depth - 1):
            queue = [child for node in queue for child in (node.left, node.right) if child]
        for node in queue:
            node.left = TreeNode(val, node.left)
            node.right = TreeNode(val, None, node.right)
        return dummy.left

File: src/623. Add One Row to Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

// DFS
// class Solution {
// public:
//     TreeNode* addOneRow(TreeNode* root, int val, int depth) {
//         if (depth == 1) return new TreeNode(val, root, NULL);
//         graft(root, val, depth);
//         return root;
//     }
//     void graft(TreeNode* root, int val, int depth) {
//         if (!root) return;
//         if (depth == 2) {
//             root->left = new TreeNode(val, root->left, NULL);
//             root->right = new TreeNode(val, NULL, root->right);
//             return;
//         }
//         graft(root->left, val, depth - 1);
//         graft(root->right, val, depth - 1);
//     }
// };

// BFS
class Solution {
public:
    TreeNode* addOneRow(TreeNode* root, int val, int depth) {
        TreeNode* dummy = new TreeNode(0, root, NULL);
        queue<TreeNode*> q;
        q.push(dummy);
        for (int i = 0; i < depth - 1; ++i) {
            for (int j = q.size(); j > 0; --j) {
                TreeNode* node = q.front();
                q.pop();
                if (node->left) q.push(node->left);
                if (node->right) q.push(node->right);
            }
        }
        while (!q.empty()) {
            TreeNode* node = q.front();
            node->left = new TreeNode(val, node->left, NULL);
            node->right = new TreeNode(val, NULL, node->right);
            q.pop();
        }
        return dummy->left;
    }
};